From 6202951896f519e50bf1d4005ee421f6e5c5d1a8 Mon Sep 17 00:00:00 2001
From: Karmjit Mahil <karmjit.mahil@igalia.com>
Date: Sat, 19 Oct 2024 13:49:09 +0200
Subject: [PATCH] tu: Add DECK_EMU to advertise being a SteamDeck

This can help running games where the Turnip driver might not be
known, preventing them from running. Some games might also make
adjustments to graphics settings if they detect the SteamDeck.

Signed-off-by: Karmjit Mahil <karmjit.mahil@igalia.com>
---
 src/freedreno/vulkan/tu_device.cc | 15 +++++++++++++++
 src/freedreno/vulkan/tu_util.cc   |  1 +
 src/freedreno/vulkan/tu_util.h    |  1 +
 3 files changed, 17 insertions(+)

diff --git a/src/freedreno/vulkan/tu_device.cc b/src/freedreno/vulkan/tu_device.cc
index fb7387d3bde6f..1be1f4bd1f2a5 100644
--- a/src/freedreno/vulkan/tu_device.cc
+++ b/src/freedreno/vulkan/tu_device.cc
@@ -899,6 +899,12 @@ tu_get_physical_device_properties_1_2(struct tu_physical_device *pdevice,
       };
    }
 
+   if (TU_DEBUG(DECK_EMU)) {
+      p->driverID = VK_DRIVER_ID_MESA_RADV;
+      memset(p->driverName, 0, sizeof(p->driverName));
+      snprintf(p->driverName, VK_MAX_DRIVER_NAME_SIZE, "radv");
+   }
+
    p->denormBehaviorIndependence =
       VK_SHADER_FLOAT_CONTROLS_INDEPENDENCE_ALL;
    p->roundingModeIndependence =
@@ -1181,6 +1187,11 @@ tu_get_properties(struct tu_physical_device *pdevice,
    props->deviceID = pdevice->dev_id.chip_id;
    props->deviceType = VK_PHYSICAL_DEVICE_TYPE_INTEGRATED_GPU;
 
+   if (TU_DEBUG(DECK_EMU)) {
+      props->vendorID = 0x1002;
+      props->deviceID = 0x163F;
+   }
+
    /* Vulkan 1.4 */
    props->dynamicRenderingLocalReadDepthStencilAttachments = true;
    props->dynamicRenderingLocalReadMultisampledAttachments = true;
@@ -1195,6 +1206,10 @@ tu_get_properties(struct tu_physical_device *pdevice,
    strcpy(props->deviceName, pdevice->name);
    memcpy(props->pipelineCacheUUID, pdevice->cache_uuid, VK_UUID_SIZE);
 
+   if (TU_DEBUG(DECK_EMU)) {
+      strcpy(props->deviceName, "AMD Custom GPU 0405 (RADV VANGOGH)");
+   }
+
    tu_get_physical_device_properties_1_1(pdevice, props);
    tu_get_physical_device_properties_1_2(pdevice, props);
    tu_get_physical_device_properties_1_3(pdevice, props);
diff --git a/src/freedreno/vulkan/tu_util.cc b/src/freedreno/vulkan/tu_util.cc
index e19d43bb8a915..9d6ed7d2e7104 100644
--- a/src/freedreno/vulkan/tu_util.cc
+++ b/src/freedreno/vulkan/tu_util.cc
@@ -56,6 +56,7 @@ static const struct debug_control tu_debug_options[] = {
    { "nofdm", TU_DEBUG_NOFDM },
    { "nocb", TU_DEBUG_NO_CONCURRENT_BINNING },
    { "forcecb", TU_DEBUG_FORCE_CONCURRENT_BINNING },
+   { "deck_emu", TU_DEBUG_DECK_EMU },
    { NULL, 0 }
 };
 
diff --git a/src/freedreno/vulkan/tu_util.h b/src/freedreno/vulkan/tu_util.h
index 7ce6d3e053af4..a867640857b34 100644
--- a/src/freedreno/vulkan/tu_util.h
+++ b/src/freedreno/vulkan/tu_util.h
@@ -75,6 +75,7 @@ enum tu_debug_flags : uint64_t
    TU_DEBUG_NOFDM                    = BITFIELD64_BIT(34),
    TU_DEBUG_NO_CONCURRENT_BINNING    = BITFIELD64_BIT(35),
    TU_DEBUG_FORCE_CONCURRENT_BINNING = BITFIELD64_BIT(36),
+   TU_DEBUG_DECK_EMU                 = BITFIELD64_BIT(37),
 };
 
 struct tu_env {
-- 
GitLab

